### ISSUE TITLE
write deadline exceeded on router connection

### ISSUE THREAD

**Issue by ripienaar** [nats-io/gnatsd#904]
hello,

gnatsd 1.4.1 stock downloaded from releases page, 64 bit linux, el6

I am experiencing some `rid:16993 - Slow Consumer Detected: WriteDeadline of 5s Exceeded` issues on above version - also 1.4.0 and same embedded.  If I simply downgrade to 1.3.latest with same configs the problem goes away.

My config:

```
listen: 0.0.0.0:5153
http: localhost:6165
max_connections: 100000
write_deadline: "5s"

cluster {
  listen: 0.0.0.0:6164

  routes = [
    nats-route://peer1:6164
    nats-route://peer2:6164
    nats-route://peer3:6164
  ]
}
```

All nodes run the same config - so each of the 3 has 3 peers including itself - this is fine because `RemoveSelfReference` should clean things up.

We have ~ 40 000 connected clients each with 5 to 10 subscriptions.  Under high fan-out we see a lot of the `rid:16993 - Slow Consumer Detected: WriteDeadline of 5s Exceeded` messages which initiates cluster route restarts and of course severely impacts available and results in message loss during the splits.

I put one of these brokers in debug mode - realistically I cannot run this network on debug it slows things down a lot etc - but I noticed a lot of `Ignoring route, already processed and sent msg` messages in the one that was in debug mode when something publishes to it, could be unrelated but thought it worth mentioning anyway

**Comment by ripienaar** [comment]
I did some tests by continuously doing huge fan outs against the entire set of connected clients - comfortably handling this without any issues on 1.3.x - 500Mbps out and 800Mbps in on the broker.  Significantly more than I need or ever see in casual use so I do not think this is a throughput problem.

icmp pings are within what you'd expect for a LAN and stable.

**Comment by kozlovic** [comment]
@ripienaar Thanks for the report. I would like to clarify something. I know that you have lots of connections and subscriptions, and in order to narrow down the search, I would like to clarify if you see the write deadline errors when the route are being established or after that, when you publish lots of messages to lots of subs across the route? Thanks!

**Comment by ripienaar** [comment]
The site is in use so there are always some traffic but I stopped most of it. It starts and then goes into a reconnect cycle as if it’s failijg during the interest propagation stage already

**Comment by ripienaar** [comment]
We can zoom later if you like

**Comment by kozlovic** [comment]
So you are saying that it happens during the interest propagation? This is strange because this code did not change between 1.3.0 and 1.4.0.. Anyway, we can chat later.

**Comment by ripienaar** [comment]
It's hard to say of course if it really is during that phase - notice level logging not enough and debug level WAY TOO MUCH :)

I can say that it either stays up till all my fleet are connected and then once I put traffic through it starts flapping or it starts flapping basically very soon after startup and never settles to a stable state.  

If that's really still during propagation is hard to say (I guess I could watch the varz output now that I think about it)

**Comment by ripienaar** [comment]
As always thank you for your time @kozlovic , for reference here's a git diff for what we had at the end of the session

```
diff --git a/server/client.go b/server/client.go
index 09563bf..b5b9446 100644
--- a/server/client.go
+++ b/server/client.go
@@ -588,10 +588,13 @@ func (c *client) flushOutbound() bool {
        // load will be to hold our lock during IO, forcing others to wait and
        // applying back pressure to the publishers sending to us.
        releaseLock := c.out.pb < maxBufSize*4
+       // releaseLock := true

        // Do NOT hold lock during actual IO unless we are behind
        if releaseLock {
                c.mu.Unlock()
+       } else {
+               c.Noticef("Not releasing the flushOutboundLock")
        }

        // flush here
@@ -1843,7 +1846,9 @@ func (c *client) closeConnection(reason ClosedState) {
                return
        }

-       c.Debugf("%s connection closed", c.typeString())
+       if c.typ == ROUTER {
+               c.Errorf("%s connection closed: %v", c.typeString(), reason)
+       }

        c.clearAuthTimer()
        c.clearPingTimer()
diff --git a/server/route.go b/server/route.go
index d258f02..7cf0853 100644
--- a/server/route.go
+++ b/server/route.go
@@ -584,8 +584,12 @@ func (s *Server) sendLocalSubsToRoute(route *client) {
        var raw [4096]*subscription
        subs := raw[:0]

+       route.Noticef("About to send notifications to route")
+
        s.sl.localSubs(&subs)

+       route.Noticef("Sending %v subscriptions", len(subs))
+
        route.mu.Lock()
        for _, sub := range subs {
                // Send SUB interest only if subject has a match in import permissions
@@ -601,7 +605,7 @@ func (s *Server) sendLocalSubsToRoute(route *client) {
        route.flushSignal()
        route.mu.Unlock()

-       route.Debugf("Sent local subscriptions to route")
+       route.Noticef("Sent local subscriptions to route")
 }

 func (s *Server) createRoute(conn net.Conn, rURL *url.URL) *client {
```

**Comment by kozlovic** [comment]
@ripienaar Thanks for your time. I will work on a branch where we will attempt to make the subscription list not trigger the condition that the fan-in fix is trying to solve. If you then can run that and verify that it works, that will be great!

**Comment by derekcollison** [comment]
Thanks to both of you. This I believe is fixed in 2.0 on master, but @kozlovic and I discussed and he will break up the initial subs to allow readLoop to run and will make sure to run under the fan-in edge condition where we start to apply a small amount of back pressure to producers.

**Comment by kozlovic** [comment]
@ripienaar Sorry I don't have a branch for you to test right now. Need to look into some things and will update you tomorrow. Thanks again.

**Comment by kozlovic** [comment]
This is addressed in commit f59cd61e672c06f9457f2a4cd2b77a9984bb1182 on the 1_4_0 branch. We did not issue a 1.4.2 release yet, but @ripienaar confirmed that he is using the branch to build the executable and therefore not blocked by the absence of the 1.4.2 release.
